[ConstraintGraph] Disallow subtype constraint contractions

xedin opened the pull request:
Presently subtype constraint is considered a candidate for contraction
via `shouldContractEdge` when left-hand side of the subtype constraint
is an inout type with type variable inside. Although it's intended to
be used only in combination with BindParam constraint assigned to the
same variable, that is actually never checked and contraction of subtype
constraint like that is invalid.

Resolves: rdar://problem/34333874

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintGraph] Disallow subtype constraint contractions

Presently subtype constraint is considered a candidate for contraction
via `shouldContractEdge` when left-hand side of the subtype constraint
is an inout type with type variable inside. Although it's intended to
be used only in combination with BindParam constraint assigned to the
same variable, that is actually never checked and contraction of subtype
constraint like that is invalid.

Resolves: rdar://problem/34333874

xedin commented:
@swift-ci please smoke test

rudkx commented on validation-test/Sema/rdar34333874.swift:
```diff
@@ -0,0 +1,24 @@
+  $0.b = 0
+}
+
+print(rdar34333874)
```
newline

xedin commented:
@swift-ci please smoke test

rudkx commented on validation-test/Sema/rdar34333874.swift:
```diff
@@ -0,0 +1,24 @@
+}
+
+let rdar34333874 = foo(C()) {
+  $0.a = 42
```
Is this really testing the same thing as the reported bug? This actually is mutating the input whereas the bug report wasn't.

xedin replied:
Example from the rdar does modify fields of `$0`

swiftix replied:
The actual original code did mutate input inside the closure. I'd suggest testing both cases to be on the safe side.

xedin replied:
Sure, sounds good, I'll add one which doesn't mutate the fields.

rudkx reviewed (commented):
Otherwise LGTM.

jrose-apple commented on validation-test/Sema/rdar34333874.swift:
```diff
@@ -0,0 +1,24 @@
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: %target-build-swift %s -o %t/a.out
+// RUN: %target-run %t/a.out
```
Drive-by comment: This should be `%target-run-simple-swift`, which gets run in more situations.

xedin replied:
Thanks, @jrose-apple ! I didn't know that existed...

xedin commented:
@swift-ci please smoke test

rudkx commented on validation-test/Sema/rdar34333874.swift:
```diff
@@ -0,0 +1,31 @@
+  return this
+}
+
+let rdar34333874 = foo(C()) {
```
Hmm, apparently I didn't submit my comment about the test. Are you sure this is testing the same thing as the original report? This appears to actually be mutating something whereas the original report wasn't.

xedin replied:
Both me and @swiftix replied to that one and I've added one more test in addition to the mutating one from SR-5861 which does mutate the fields of the argument.

xedin commented:
I've fixed a problem in `isStrictInoutSubtypeConstraint` which returned true only if the left-hand side of the subtype with inout is not a type-variable, which allowed bind params to contracted incorrectly and added one more test with a single statement closure.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility
